Model rounding mode control for floating-point intrinsics

eopXD opened the pull request:
NOTE: The discussion under #222 has not converged to a finalized scheme for how we let users specify the rounding mode, so this is still a draft PR.

This PR resembles #222 and aims to model rounding mode (`frm`) control for the floating-point intrinsics.

topperc commented on auto-generated/llvm-overloaded-tests/vfncvt_rod.c:
```diff
@@ -78,3 +78,363 @@ vfloat32m4_t test_vfncvt_rod_f_f_w_f32m4_m(vbool8_t mask, vfloat64m8_t src, size
+  return __riscv_vfncvt_rod_f(mask, src, vl);
+}
+
+vfloat16mf4_t test_vfncvt_rod_f_f_w_f16mf4(vfloat32mf2_t src, size_t vl) {
```
This function name is already on line 81.

topperc commented on auto-generated/api-testing/vfrsqrt7.c:
```diff
@@ -124,3 +124,603 @@ vfloat64m8_t test_vfrsqrt7_v_f64m8_m(vbool8_t mask, vfloat64m8_t op1, size_t vl)
   return __riscv_vfrsqrt7_v_f64m8_m(mask, op1, vl);
 }
 
+vfloat16mf4_t test_vfrsqrt7_v_f16mf4_rne(vfloat16mf4_t op1, size_t vl) {
```
vfrsqrt doesn't use frm

topperc commented on auto-generated/api-testing/vwcvt.c:
```diff
@@ -124,3 +124,603 @@ vint64m8_t test_vwcvt_x_x_v_i64m8_m(vbool8_t mask, vint32m4_t src, size_t vl) {
   return __riscv_vwcvt_x_x_v_i64m8_m(mask, src, vl);
 }
 
+vint16mf4_t test_vwcvt_x_x_v_i16mf4_rne(vint8mf8_t src, size_t vl) {
```
this is an integer instruction

topperc commented on auto-generated/llvm-api-tests/vfwcvt.c:
```diff
@@ -462,3 +462,2283 @@ vfloat64m8_t test_vfwcvt_f_f_v_f64m8_m(vbool8_t mask, vfloat32m4_t src, size_t v
+  return __riscv_vfwcvt_f_f_v_f64m8_m(mask, src, vl);
+}
+
+vfloat16mf4_t test_vfwcvt_f_x_v_f16mf4(vint8mf8_t src, size_t vl) {
```
This test name is repeated

topperc commented on auto-generated/llvm-overloaded-tests/vfneg.c:
```diff
@@ -126,3 +126,603 @@ vfloat64m8_t test_vfneg_v_f64m8_m(vbool8_t mask, vfloat64m8_t op1, size_t vl) {
   return __riscv_vfneg(mask, op1, vl);
 }
 
+vfloat16mf4_t test_vfneg_v_f16mf4_rne(vfloat16mf4_t op1, size_t vl) {
```
fneg doesn't use frm

topperc commented on auto-generated/api-testing/vfwcvt.c:
```diff
@@ -460,3 +460,2283 @@ vfloat64m8_t test_vfwcvt_f_f_v_f64m8_m(vbool8_t mask, vfloat32m4_t src, size_t v
+  return __riscv_vfwcvt_f_xu_v_f16m8(src, vl);
+}
+
+vint32mf2_t test_vfwcvt_x_f_v_i32mf2(vfloat16mf4_t src, size_t vl) {
```
Should this have a rounding mode? Float to int loses decimal places which need to be rounded.

eopXD replied:
Yes I think so. Thank you for catching the miss.

zhongjuzhe commented:
conversions with 'rtz' should not have FRM.

topperc commented on auto-generated/api-testing/vfadd.c:
```diff
@@ -244,3 +244,243 @@ vfloat64m8_t test_vfadd_vf_f64m8_m(vbool8_t mask, vfloat64m8_t op1, float64_t op
 }
 
+vfloat16mf4_t test_vfadd_vv_f16mf4_rm(vfloat16mf4_t op1, vfloat16mf4_t op2, size_t vl) {
+  return __riscv_vfadd_vv_f16mf4_rm(op1, op2, FRM_RNE, vl);
```
We can't use `FRM_RNE`. It needs to be prefixed with underscores. Probably needs to be `__FRM_RNE` or `__RISCV_FRM_RNE`

eopXD replied:
Updated the enum to 

```c
enum __RISCV_FRM {
  __RISCV_FRM_RNE = 0,
  __RISCV_FRM_RTZ = 1,
  __RISCV_FRM_RDN = 2,
  __RISCV_FRM_RUP = 3,
  __RISCV_FRM_RMM = 4,
};

```

topperc commented on auto-generated/api-testing/vfcvt.c:
```diff
@@ -484,3 +484,483 @@ vfloat64m8_t test_vfcvt_f_xu_v_f64m8_m(vbool8_t mask, vuint64m8_t src, size_t vl
 }
 
+vint16mf4_t test_vfcvt_x_f_v_i16mf4_rm(vfloat16mf4_t src, size_t vl) {
+  return __riscv_vfcvt_x_f_v_i16mf4_rm(src, vl);
```
This says _rm but there's no rounding mode being passed

topperc commented on auto-generated/api-testing/vfrec7.c:
```diff
@@ -124,3 +124,123 @@ vfloat64m8_t test_vfrec7_v_f64m8_m(vbool8_t mask, vfloat64m8_t op1, size_t vl) {
 }
 
+vfloat16mf4_t test_vfrec7_v_f16mf4_rm(vfloat16mf4_t op1, size_t vl) {
+  return __riscv_vfrec7_v_f16mf4_rm(op1, vl);
```
No rounding mode being passed

topperc commented on auto-generated/api-testing/vfredmax.c:
```diff
@@ -124,3 +124,123 @@ vfloat64m1_t test_vfredmax_vs_f64m8_f64m1_m(vbool8_t mask, vfloat64m8_t vector,
 }
 
+vfloat16m1_t test_vfredmax_vs_f16mf4_f16m1_rm(vfloat16mf4_t vector, vfloat16m1_t scalar, size_t vl) {
+  return __riscv_vfredmax_vs_f16mf4_f16m1_rm(vector, scalar, vl);
```
No rounding mode

topperc commented on auto-generated/api-testing/vfredosum.c:
```diff
@@ -124,3 +124,123 @@ vfloat64m1_t test_vfredosum_vs_f64m8_f64m1_m(vbool8_t mask, vfloat64m8_t vector,
 }
 
+vfloat16m1_t test_vfredosum_vs_f16mf4_f16m1_rm(vfloat16mf4_t vector, vfloat16m1_t scalar, size_t vl) {
+  return __riscv_vfredosum_vs_f16mf4_f16m1_rm(vector, scalar, vl);
```
No rounding mode

topperc commented on auto-generated/api-testing/vfwcvt.c:
```diff
@@ -460,3 +460,459 @@ vfloat64m8_t test_vfwcvt_f_f_v_f64m8_m(vbool8_t mask, vfloat32m4_t src, size_t v
 }
 
+vfloat16mf4_t test_vfwcvt_f_x_v_f16mf4_rm(vint8mf8_t src, size_t vl) {
+  return __riscv_vfwcvt_f_x_v_f16mf4_rm(src, vl);
```
No rounding mode

topperc commented on auto-generated/api-testing/vfwredosum.c:
```diff
@@ -92,3 +92,91 @@ vfloat64m1_t test_vfwredosum_vs_f32m8_f64m1_m(vbool4_t mask, vfloat32m8_t vector
 }
 
+vfloat32m1_t test_vfwredosum_vs_f16mf4_f32m1_rm(vfloat16mf4_t vector, vfloat32m1_t scalar, size_t vl) {
+  return __riscv_vfwredosum_vs_f16mf4_f32m1_rm(vector, scalar, vl);
```
No rounding mode

topperc commented on auto-generated/api-testing/vwcvtu.c:
```diff
@@ -124,3 +124,123 @@ vuint64m8_t test_vwcvtu_x_x_v_u64m8_m(vbool8_t mask, vuint32m4_t src, size_t vl)
 }
 
+vuint16mf4_t test_vwcvtu_x_x_v_u16mf4_rm(vuint8mf8_t src, size_t vl) {
+  return __riscv_vwcvtu_x_x_v_u16mf4_rm(src, vl);
```
This an integer to integer conversion. It's not FP

zixuan-wu commented:
I am wondering if we need a dynamic mode enum which representing using current vxrm value? @eopXD

topperc commented on auto-generated/api-testing/vfadd.c:
```diff
@@ -244,3 +244,243 @@ vfloat64m8_t test_vfadd_vf_f64m8_m(vbool8_t mask, vfloat64m8_t op1, float64_t op
 }
 
+vfloat16mf4_t test_vfadd_vv_f16mf4_rm(vfloat16mf4_t op1, vfloat16mf4_t op2, size_t vl) {
+  return __riscv_vfadd_vv_f16mf4_rm(op1, op2, __RISCV_FRM_RNE, vl);
```
Are the __RISCV_FRM_RNE and other values documented? The web view makes it hard to find if it is.

kito-cheng replied:
Changes within this repo generally will break into at least two part: 1) generator/doc update and 2) re-run generator.

So personally I will switch to the `commit` tab and found those doc change for easier review :)

e.g.
https://github.com/riscv-non-isa/rvv-intrinsic-doc/pull/226/commits/f1ad3258b9e7de52f57c811cf90d3b5541604f94

dzaima commented:
Is there a reason the `_rm` suffix is after the policy suffix? Currently, as far as I know, the policy suffix is always the at the end, which allows automatically transforming (i.e. via C macro concatenation or otherwise) any instruction to one with specific policy via appending to it, but this would introduce the need to sometimes put it in the middle instead.

In my opinion, it would be better to have `_rm` before the policy, i.e. `__riscv_vfadd_vv_f32m1_rm_m`, such that `__riscv_vfadd_vv_f32m1_rm` by itself is a regular intrinsic that can be masked by appending to its name.

eopXD commented:
> Is there a reason the `_rm` suffix is after the policy suffix? Currently, as far as I know, the policy suffix is always the at the end, which allows automatically transforming (i.e. via C macro concatenation or otherwise) any instruction to one with specific policy via appending to it, but this would introduce the need to sometimes put it in the middle instead.
> 
> In my opinion, it would be better to have `_rm` before the policy, i.e. `__riscv_vfadd_vv_f32m1_rm_m`, such that `__riscv_vfadd_vv_f32m1_rm` by itself is a regular intrinsic that can be masked by appending to its name.

It was more natural to me when I created the PR that the `_rm` is appending "after" the policy suffix because we are keeping the existing ones with semantic that does not alternate the rounding mode and use what is already within `frm`.

However I think you made a valid point, I will update the PR.

---

On the other hand, I see your good work https://dzaima.github.io/intrinsics-viewer/ that will be helpful for the RVV intrinsics users. I sent an email that is linked with your Github commits <dzaimagit [at] gmail.com>, please reply to me if possible, thank you.

eopXD commented:
Rebased to master.

eopXD commented:
A short summary on the updates. The description at the top of this PR was also updated.

The first version of the proposal proposes to append the suffix of the rounding mode into the function name, adding variants like the following which is taking unmasked fadd intrinsics as an example.

```cpp
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rne (vfloat32m1_t op1, vfloat32m1_t op2, size_t vl);
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rtz (vfloat32m1_t op1, vfloat32m1_t op2, size_t vl);
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rdn (vfloat32m1_t op1, vfloat32m1_t op2, size_t vl);
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rup (vfloat32m1_t op1, vfloat32m1_t op2, size_t vl);
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rmm (vfloat32m1_t op1, vfloat32m1_t op2, size_t vl);
```

---

Then as suggested by Craig, the proposal changed to having the rounding mode control as parameter instead of appended into the suffix. The variant with rounding mode control is appended after the policy suffixes (if any). Using the `_mu` variant as an example here.

```cpp
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rm (vfloat32m1_t op1, vfloat32m1_t op2, unsigned int frm, size_t vl);
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_mu_rm (vbool32_t mask, vfloat32m1_t maskedoff, vfloat32m1_t op1, vfloat32m1_t op2, unsigned int frm, size_t vl);
```

---

In the latest force-push, adjusted upon dzaima's comment, the proposal now appends the `_rm` suffix before the policy suffix, so it could be obvious that the `_rm` variant also provides a whole set of policy variants. Using the `_mu` variant as an example here.

```cpp
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rm (vfloat32m1_t op1, vfloat32m1_t op2, unsigned int frm, size_t vl);
+ vfloat32m1_t __riscv_vfadd_vv_f32m1_rm_mu (vbool32_t mask, vfloat32m1_t maskedoff, vfloat32m1_t op1, vfloat32m1_t op2, unsigned int frm, size_t vl);
```

Incarnation-p-lee commented on auto-generated/api-testing/vfadd.c:
```diff
@@ -244,3 +244,243 @@ vfloat64m8_t test_vfadd_vf_f64m8_m(vbool8_t mask, vfloat64m8_t op1, float64_t op
+}
+
+vfloat32m1_t test_vfadd_vv_f32m1_rm_m(vbool32_t mask, vfloat32m1_t op1, vfloat32m1_t op2, size_t vl) {
+  return __riscv_vfadd_vv_f32m1_rm_m(mask, op1, op2, __RISCV_FRM_RNE, vl);
```
should be `__riscv_vfadd_vv_f32m1_m_rm` instead of `__riscv_vfadd_vv_f32m1_rm_m`?

eopXD replied:
We are having `_rm_m` for the symmetry of `__riscv_vfadd_vv_f32m1` and `__riscv_vfadd_vv_f32m1_rm` as two sets (the implicit-`frm` intrinsics and the explicit-`frm` intrinsics), and then the policy variants be applied upon these two sets.

eopXD replied:
Just updated the description, thank you for noticing.

Incarnation-p-lee replied:
Cool, thank you for confirmation.
